### PR TITLE
Added .NET Core Support (OpenNLP.NET 1.9.1.2-preview0001)

### PR DESCRIPTION
**Pull request by NightOwl888**
- Added netcoreapp3.1 target
- Added tests for netcoreapp3.1
- Bumped from net45 to net461
- Upgraded to IKVM [8.2.0-prerelease0392](https://www.nuget.org/packages/IKVM/8.2.0-prerelease0392)
- Allow building with any .NET SDK over 6.x

This relies on a preview build of IKVM, the first release with .NET Core support (as well as .NET 5+). This IKVM build is _definitely not production ready_, but we could use some help with testing to get it ready, so it would help if this is released as a preview on NuGet when you get a chance.

I have made a build of OpenNLP.NET and run our 26 tests on macos-latest, ubuntu-latest, and windows-latest (the latter both in x64 and x86) on net48, netcoreapp3.1, net5.0, and net6.0 and they are all green. The 6 OpenNLP.NET tests are also passing on windows for both net461 and netcoreapp3.1, but I am having trouble getting Paket to download anything on macOS in GitHub Actions. I attempted to upgrade Paket, but that didn't change the results.

Do note that there is no official `ikvmc` version to build from macOS at this point, since the tools rely on native bits that haven't been ported yet to macOS. And it will probably be dumbed down in the future to emulate the way csc.dll works on .NET Core so the current command options may break in a future `ikvmc` release. I changed the strategy for GitHub Actions to build on Windows and then transfer the binaries to Windows, Linux and macOS to test on.

Also note that the assemblies don't have a [TargetFrameworkAttribute](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.versioning.targetframeworkattribute?view=net-6.0) so when viewing them in decompilers, the .NET Core assemblies show as .NET Framework 4.0. If you wish to add the attribute, you will need to create `.java` files that contain annotations using [this syntax](https://github.com/ikvm-revived/ikvm/blob/2d2d47946006d302de8f2db2ade24d3e5f0ebb8b/openjdk/AssemblyInfo.java.in#L30) and then reference the file with the [`-assemblyattributes`](https://github.com/ikvm-revived/ikvm/blob/acc33f8a88abc21eb18c2706bee6e6118eb66d0d/ikvmc/IkvmcCompiler.cs#L386-L387) option in `ikvmc`.

There is a bug in Fake that is preventing from releasing a lower version than the last release during the release notes parsing (https://github.com/fsprojects/FAKE/issues/2680), so I added https://github.com/sergey-tihon/OpenNLP.NET/commit/3ba44b58b3e447cec41c53682cf7acbc3a5c6c7f which can be reverted after the release to put the 1.9.3 and 1.9.4 releases back in. I will leave it up to you whether you want to upgrade those now as previews, or wait until IKVM has a stable release.

### CODE REVIEW

**Comment by NightOwl888** [comment]
Alright, I got all of the tests passing on Windows, macOS and Linux in GitHub Actions after adding a separate stage for testing.

There are still some things that could be optimized.  But, since I have limited experience with Paket and Fake, I would prefer that you tackled them, if interested. The entire build/test takes only 15 minutes now, and the optimizations may save about 1/3 of that time.

1. Dependencies can be pinned to specific frameworks (when I tried it, it failed but maybe there is some trick to it that is beyond my (limited) understanding of Paket)
2. Model binaries could be embedded resources in the test DLL, so they wouldn't have to be downloaded on each test agent separately.
3. I don't think that .NET Core 3.1 is required to build, but it is for testing.

**Comment by sergey-tihon** [comment]
Thank you @NightOwl888! IKVM for .NET Core is a really awesome news.

Do you have use case for targeting `netcoreapp3.1`? It is close to its end of life already.
I personally would prefer to go with `net6.0` simplify maintenance and life of future generations ;)

**Comment by NightOwl888** [comment]
> Do you have use case for targeting `netcoreapp3.1`? It is close to its end of life already.
> I personally would prefer to go with `net6.0` simplify maintenance and life of future generations ;)

Currently, `netcoreapp3.1` is the only target that works with `ikvmc`. But consuming applications can use any higher version, such as .NET 5, .NET 6, or .NET 7 preview (although the last case hasn't been tested).

image

The future plans are to keep `ikvmc` but dumb it down like `csc.dll` in .NET Core, then we will build a higher level tool such as `ikvm build` that allows to select a target. You would be able to select `.net6.0` even if the compiler were targeting `netcoreapp3.1`, you would just need to point it to the reference assemblies for `.net6.0`. But the code is still too much of a mess to do that dynamically because of all of the hard-coded #ifdef statements.

Under the hood in .NET, there are actually only 4 different assembly versions and all of .NET Core runs on the same version. This is why I mentioned the `TargetFrameworkAttribute` above - it can be put into a template that can be added to the `ikvmc` command so .NET decompilers (such as dotPeek) will see the actual target framework instead of ".NET Framework 4.0", which is what it shows now.

image

Here is an example of an assembly compiled with `dotnet build`:

image

But the current thinking will be to discourage putting IKVM "ports" like this on NuGet.org and instead package `.targets` files with [`<MavenReference>` elements](https://github.com/ikvm-revived/ikvm/issues/50) in the `tools` directory of a `.nupkg`, then have the build tools compile a private build for each consumer. I am skeptical of this approach, but at the same time I can't seem to poke any holes in it. If a consumer references multiple packages that have the same `<MavenReference>` elements they will still end up with only a single copy of the binaries based on the Maven package version resolution, which will be compiled locally and either distributed with an executable or re-packaged as a `.targets` file only with no DLLs in a consuming `.nupkg` when using `dotnet pack`.

For the time being, since OpenNLP.NET is already in the wild, it makes a good test case until the IKVM tooling is completed.

**Comment by sergey-tihon** [comment]
> BTW - I am curious where you came up with [this approach](https://github.com/sergey-tihon/OpenNLP.NET/blob/6229220a659d6ed83df13d251364fa5fc7e79cd0/build.fsx#L140-L144) for writing the strong name key as a BLOB.


I was looking a way to sign unsigned assemblies restored from NuGet, because we had to deploy in GAC. Back in a day I found that it is possible to do using [Mono.Cecil](https://www.nuget.org/packages/Mono.Cecil/) by Jb Evans using [StrongNaming capabilities](https://github.com/jbevain/cecil/wiki/StrongName)

In 2019 during adoption to .NET Core 2.0 they changed API https://github.com/jbevain/cecil/pull/548 and introduced option to load key from BLOB
https://github.com/jbevain/cecil/blob/79b43e8e72866f450dee8b59510cb5767b1491ec/Mono.Security.Cryptography/CryptoService.cs#L177-L178

File: paket.dependencies [hunk 1]
```diff
@@ -1,13 +1,17 @@
...
+nuget Microsoft.NET.Test.Sdk
+nuget Microsoft.NETFramework.ReferenceAssemblies
+
+http https://github.com/ikvm-revived/ikvm/releases/download/8.2.0-prerelease.392/IKVM-8.2.0-prerelease.392-tools-net461-any.zip ikvm-8.2.0-prerelease.392-tools-net461-win7-x64/tools.zip
+http https://github.com/ikvm-revived/ikvm/releases/download/8.2.0-prerelease.392/IKVM-8.2.0-prerelease.392-tools-netcoreapp3.1-win7-x64.zip ikvm-8.2.0-prerelease.392-tools-netcoreapp3.1-linux-x64/tools.zip
+http https://github.com/ikvm-revived/ikvm/releases/download/8.2.0-prerelease.392/IKVM-8.2.0-prerelease.392-tools-netcoreapp3.1-linux-x64.zip ikvm-8.2.0-prerelease.392-tools-netcoreapp3.1-win7-x64/tools.zip
 
-http https://sergeytihon.files.wordpress.com/2021/11/ikvmbin-8.1.5717.0.zip
-http https://dlcdn.apache.org/opennlp/opennlp-1.9.4/apache-opennlp-1.9.4-bin.zip
+http https://archive.apache.org/dist/opennlp/opennlp-1.9.1/apache-opennlp-1.9.1-bin.zip
```
Comment (sergey-tihon) [inline paket.dependencies:14]: Why you reverted OpenNLP from `1.9.4` to `1.9.1`?

**Comment by sergey-tihon** [comment]
> but we could use some help with testing to get it ready, so it would help if this is released as a preview on NuGet when you get a chance.

Done, package available on NuGet - https://www.nuget.org/packages/OpenNLP.NET/1.9.1.2-preview0001

>  I am having trouble getting Paket to download anything on macOS in GitHub Actions. I attempted to upgrade Paket, but that didn't change the results.

This is looks resolved for me

> Do note that there is no official ikvmc version to build from macOS at this point

This is a little bit unfortunate for me personally, because I work primarily on macOS these days and have hard access to Windows.

> package .targets files with https://github.com/ikvm-revived/ikvm/issues/50 in the tools directory of a .nupkg, then have the build tools compile a private build for each consumer.

I am also skeptical, it sounds promising but I am not sure that it will work technically. How would you determine dependencies between `jar` files and build order?

In this repo, I've tried to use [jdeps](https://github.com/sergey-tihon/OpenNLP.NET/blob/master/build.fsx#L149-L158) to infer dependencies between `jar` files, export them as a graph in `*.dot` format and then use that graph to recompile them with `ikvmc`.

But this approach failed on [Stanford.NLP.NET)](https://github.com/sergey-tihon/Stanford.NLP.NET/) and I still [manually maintain](https://github.com/sergey-tihon/Stanford.NLP.NET/blob/master/build.fsx#L218-L242) dependencies, versions and build order. IIRC, `jdeps` just failed on folder with Stanford jars

image

File: paket.dependencies (hunk 1 above)
Comment (NightOwl888) [reply paket.dependencies:14]: This is done because Lucene 8.2.0 (the version which we have ported our OpenNLP extension from) [targeted 1.9.1](https://github.com/apache/lucene/blob/releases/lucene-solr/8.2.0/lucene/ivy-versions.properties#L181). The most recent version, 9.2.0, also still [targets 1.9.1](https://github.com/apache/lucene/blob/releases/lucene/9.2.0/versions.props#L14).

In theory, it should work with 1.9.4 if they are using semantic versioning. This is just to mitigate any risk where the version makes a difference.

**Comment by NightOwl888** [comment]
> > but we could use some help with testing to get it ready, so it would help if this is released as a preview on NuGet when you get a chance.
> 
> Done, package available on NuGet - https://www.nuget.org/packages/OpenNLP.NET/1.9.1.2-preview0001

Thanks. It looks like we ended up pushing an update to IKVM to fix some bugs that were reported with resolving class paths.

https://www.nuget.org/packages/IKVM/8.2.0-prerelease0809

I don't have any direct evidence that this patch directly affects OpenNLP APIs, but we have a very small sampling of the tests for OpenNLP, so there could be some parts of it that are not functional without the IKVM patch (and even with it).

> > I am having trouble getting Paket to download anything on macOS in GitHub Actions. I attempted to upgrade Paket, but that didn't change the results.
> 
> This is looks resolved for me

Yes, sorry that comment was already obsolete before I posted the PR.

> > Do note that there is no official ikvmc version to build from macOS at this point
> 
> This is a little bit unfortunate for me personally, because I work primarily on macOS these days and have hard access to Windows.

Actually, it will probably function with the Windows or Linux builds if you run `dotnet ikvmc.dll` from one of those distributions. The only thing that probably won't work is if the Java app depends on the native bits.

> > package .targets files with [ikvm-revived/ikvm#50](https://github.com/ikvm-revived/ikvm/issues/50) in the tools directory of a .nupkg, then have the build tools compile a private build for each consumer.
> 
> I am also skeptical, it sounds promising but I am not sure that it will work technically. How would you determine dependencies between `jar` files and build order?
> 
> In this repo, I've tried to use [jdeps](https://github.com/sergey-tihon/OpenNLP.NET/blob/master/build.fsx#L149-L158) to infer dependencies between `jar` files, export them as a graph in `*.dot` format and then use that graph to recompile them with `ikvmc`.
> 
> But this approach failed on [Stanford.NLP.NET)](https://github.com/sergey-tihon/Stanford.NLP.NET/) and I still [manually maintain](https://github.com/sergey-tihon/Stanford.NLP.NET/blob/master/build.fsx#L218-L242) dependencies, versions and build order. IIRC, `jdeps` just failed on folder with Stanford jars
> 
> image

Good info to have. We will definitely need to find an approach that works across Java build tools. Also dependency resolution should work if there are mixed [`<JavaReference>`](https://github.com/ikvm-revived/ikvm/issues/54) and [`<MavenReference>`](https://github.com/ikvm-revived/ikvm/issues/50) elements in a single project.

`<MavenReference>` is high-level functionality and there are several steps we need to complete before getting to that design, though.

I am curious what the limitation of `jdeps` was that prevented it from working on Standford.NLP.NET? Perhaps this is something that would be fixed in a newer version of Java?

Looks like an alternative is [JarAnalyzer](https://mvnrepository.com/artifact/com.kirkk/jaranalyzer/1.2). Although, the original source URL is offline, [it is still available on the Internet Archive](https://web.archive.org/web/20161014074817/http://www.kirkk.com/main/Main/JarAnalyzer).

**Comment by sergey-tihon** [comment]
> I am curious what the limitation of jdeps was that prevented it from working on Standford.NLP.NET? Perhaps this is something that would be fixed in a newer version of Java?

I believe this one -  https://issues.apache.org/jira/browse/MJDEPS-15

When I run it on all assemblies I get an error that there are multi-release packages
```
jaxb-api-2.4.0-b180830.0359.jar is a multi-release jar file but --multi-release option is not set
```
and with `--multi-release` it fails on non-multirelease packages

So caller should identify [Multi-Release JAR Files](https://openjdk.java.net/jeps/238) manually (at least now) and process them separately

**Comment by NightOwl888** [comment]
You might want to check out the [latest work on `<MavenRefernece>` support](https://github.com/ikvm-revived/MavenReferencePoC). We got it to use Maven to download all 37 `.jar` files for this project. And it is set up to allow a project-level `pom.xml` to configure private Maven repositories and credentials. Actually, there is just one `.targets` file that does all the work, which could be imported into a project manually. Although, this approach probably won't work if you want to stick with Paket and Fake.

**Comment by sergey-tihon** [comment]
@NightOwl888 Am I right that in order to build `netcore` assemblies from `*.jar` I need to tun `ikvmc` from `KVM-8.2.0-prerelease.392-tools-netcoreapp3.1-win7-x64.zip`?

I found strange code in this PR
- These lines swap `win7` and `linux` binaries during the download - https://github.com/sergey-tihon/OpenNLP.NET/pull/14/files#diff-377fe73bf860b37e7d9a8ca5be27762668b3a08aac581b435d4c31b737d905cdR11-R12
- and then here, you always use `net461` for both target frameworks - https://github.com/sergey-tihon/OpenNLP.NET/pull/14/files#diff-5ebd60ceb88910043c04c5a665d836197d7184488edcf52f590e8fcad6101519R85-R86


With such build process it is not clear what the point to build for 2 targets `net461` and `netcoreapp3.1` ...

**Comment by NightOwl888** [comment]
> @NightOwl888 Am I right that in order to build `netcore` assemblies from `*.jar` I need to tun `ikvmc` from `KVM-8.2.0-prerelease.392-tools-netcoreapp3.1-win7-x64.zip`?
> 
> I found strange code in this PR
> 
> * These lines swap `win7` and `linux` binaries during the download - https://github.com/sergey-tihon/OpenNLP.NET/pull/14/files#diff-377fe73bf860b37e7d9a8ca5be27762668b3a08aac581b435d4c31b737d905cdR11-R12
> * and then here, you always use `net461` for both target frameworks - https://github.com/sergey-tihon/OpenNLP.NET/pull/14/files#diff-5ebd60ceb88910043c04c5a665d836197d7184488edcf52f590e8fcad6101519R85-R86
> 
> With such build process it is not clear what the point to build for 2 targets `net461` and `netcoreapp3.1` ...

You are right. These were not meant to be swapped around. Unfortunately, not having a combined .NET Core executable is a bit confusing.

**Comment by sergey-tihon** [comment]
I've tried to update to the next release of IKVM and run `netcoreapp3.1` build using `netcoreapp3.1` version of IKVMC.exe

```sharp
let ikvmVersion = @"8.2.0-prerelease.809"
let ikvmRootFolder = root </> "paket-files" </> "github.com"

let ikvmcFolder_NetFramework = ikvmRootFolder </> "ikvm-" + ikvmVersion + "-tools-net461-win7-x64"
let ikvmcFolder_NetCore_Windows = ikvmRootFolder </> "ikvm-" + ikvmVersion + "-tools-netcoreapp3.1-win7-x64"
```

and it [fails](https://github.com/sergey-tihon/OpenNLP.NET/runs/6849832947?check_suite_focus=true) with 
```
*** INTERNAL COMPILER ERROR ***
[676](https://github.com/sergey-tihon/OpenNLP.NET/runs/6849832947?check_suite_focus=true#step:7:677)
[677](https://github.com/sergey-tihon/OpenNLP.NET/runs/6849832947?check_suite_focus=true#step:7:678)
PLEASE FILE A BUG REPORT FOR IKVM.NET WHEN YOU SEE THIS MESSAGE
[678](https://github.com/sergey-tihon/OpenNLP.NET/runs/6849832947?check_suite_focus=true#step:7:679)
[679](https://github.com/sergey-tihon/OpenNLP.NET/runs/6849832947?check_suite_focus=true#step:7:680)
ikvmc, Version=8.2.0.0, Culture=neutral, PublicKeyToken=13235d27fcbfff58
[680](https://github.com/sergey-tihon/OpenNLP.NET/runs/6849832947?check_suite_focus=true#step:7:681)
D:\a\OpenNLP.NET\OpenNLP.NET\paket-files\github.com\ikvm-8.2.0-prerelease.809-tools-netcoreapp3.1-win7-x64\
[681](https://github.com/sergey-tihon/OpenNLP.NET/runs/6849832947?check_suite_focus=true#step:7:682)
3.1.25 64-bit
[682](https://github.com/sergey-tihon/OpenNLP.NET/runs/6849832947?check_suite_focus=true#step:7:683)
[683](https://github.com/sergey-tihon/OpenNLP.NET/runs/6849832947?check_suite_focus=true#step:7:684)
System.IO.FileNotFoundException: Reference assemblies directory: D:\a\OpenNLP.NET\OpenNLP.NET\packs\github.com.Ref\3.1.0\ref\netcoreapp3.1
[684](https://github.com/sergey-tihon/OpenNLP.NET/runs/6849832947?check_suite_focus=true#step:7:685)
   at IKVM.Reflection.Universe.BuildRefDirFrom(String runtimeDir) in D:\a\ikvm\ikvm\IKVM.Reflection\Universe.cs:line 180
[685](https://github.com/sergey-tihon/OpenNLP.NET/runs/6849832947?check_suite_focus=true#step:7:686)
   at IKVM.Reflection.Universe.get_ReferenceAssembliesDirectory() in D:\a\ikvm\ikvm\IKVM.Reflection\Universe.cs:line 156
[686](https://github.com/sergey-tihon/OpenNLP.NET/runs/6849832947?check_suite_focus=true#step:7:687)
   at IKVM.Reflection.Universe.DefaultResolver(String refname, Boolean throwOnError) in D:\a\ikvm\ikvm\IKVM.Reflection\Universe.cs:line 842
[687](https://github.com/sergey-tihon/OpenNLP.NET/runs/6849832947?check_suite_focus=true#step:7:688)
   at IKVM.Reflection.Universe.Load(String refname, Module requestingModule, Boolean throwOnError) in D:\a\ikvm\ikvm\IKVM.Reflection\Universe.cs:line 799
[688](https://github.com/sergey-tihon/OpenNLP.NET/runs/6849832947?check_suite_focus=true#step:7:689)
   at IKVM.Internal.AssemblyResolver.Init(Universe universe, Boolean nostdlib, IList`1 references, IList`1 userLibPaths) in D:\a\ikvm\ikvm\ikvmc\IKVM\Internal\AssemblyResolver.cs:line 93
[689](https://github.com/sergey-tihon/OpenNLP.NET/runs/6849832947?check_suite_focus=true#step:7:690)
   at ikvmc.IkvmcCompiler.Compile(String[] args) in D:\a\ikvm\ikvm\ikvmc\IkvmcCompiler.cs:line 178
[690](https://github.com/sergey-tihon/OpenNLP.NET/runs/6849832947?check_suite_focus=true#step:7:691)
   at ikvmc.IkvmcCompiler.Main(String[] args) in D:\a\ikvm\ikvm\ikvmc\IkvmcCompiler.cs:line 112
[691](https://github.com/sergey-tihon/OpenNLP.NET/runs/6849832947?check_suite_focus=true#step:7:692)
Finished (Failed) 'Compile' in 00:01:14.7279118
```

currently I am not sure what is the right way to use `ikvmc` and build assemblies for `netcore`

**Comment by NightOwl888** [comment]
> I've tried to update to the next release of IKVM and run `netcoreapp3.1` build using `netcoreapp3.1` version of IKVMC.exe
> 
> ```
> let ikvmVersion = @"8.2.0-prerelease.809"
> let ikvmRootFolder = root </> "paket-files" </> "github.com"
> 
> let ikvmcFolder_NetFramework = ikvmRootFolder </> "ikvm-" + ikvmVersion + "-tools-net461-win7-x64"
> let ikvmcFolder_NetCore_Windows = ikvmRootFolder </> "ikvm-" + ikvmVersion + "-tools-netcoreapp3.1-win7-x64"
> ```
> 
> and it [fails](https://github.com/sergey-tihon/OpenNLP.NET/runs/6849832947?check_suite_focus=true) with
> 
> image
> 
> currently I am not sure what is the right way to use `ikvmc` and build assemblies for `netcore`

You probably need to use `-nostdlib -r:<ikvmtools>/refs/*.dll` as mentioned [here](https://github.com/ikvm-revived/ikvm/issues/79#issuecomment-1152234536). Unfortunately, .NET Core is not finding its reference assemblies automatically.

**Comment by sergey-tihon** [comment]
Close this PR because https://github.com/sergey-tihon/OpenNLP.NET/pull/15 is merged.
